Add configurable HTTP retry mechanism for OCSP validation

chance-coleman opened the pull request:
## Description
closes #42401 

This PR adds two key improvements to Keycloak:

1. **General-Purpose HTTP Retry Mechanism**: Implements a general-purpose HTTP retry mechanism that allows automatic retrying of HTTP requests when they fail due to transient issues such as network timeouts, temporary server errors, or other recoverable conditions.

2. **OCSP Retry UI Configuration**: Adds UI configuration inputs for OCSP retry count and timeout settings in the X.509 client certificate authentication flow. This allows administrators to configure retry behavior for OCSP certificate validation through the Keycloak admin UI.

The implementation extends the existing `HttpClientProvider` interface to provide retriable HTTP clients that can be used throughout the codebase. This approach provides a consistent way to handle retries across different components, including OCSP certificate validation.

## Motivation

In environments with unreliable network connections or when interacting with services that may experience temporary outages, HTTP requests can fail unnecessarily. This is particularly problematic for critical operations like OCSP certificate validation, where a temporary network issue could prevent a user from authenticating.

By implementing an opt-in general-purpose retry mechanism, we can improve the reliability of Keycloak in these environments without changing default behavior or duplicating retry logic across different components.

chance-coleman committed:
Add OCSP retry configuration and HTTP retry mechanism
Implements a configurable retry mechanism for OCSP certificate validation with UI configuration in the X.509 client certificate authenticator.

Also adds a general-purpose HTTP retry mechanism with exponential backoff and jitter that can be used throughout the codebase.

The implementation is opt-in by default (0 retries) to maintain backward compatibility while allowing configuration when needed.

Closes #42401

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

cleanup

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

cleanup

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

cleanup

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

cleanup

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

chance-coleman committed:
refactor retriable client to use cached cclient to be more optimized

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

slaskawi reviewed (approved):
LGTM with a small nitpick, which is up to you to implement.

chance-coleman committed:
implement equals override to simplify equality of httpclients

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

rmartinc reviewed (commented):
Thanks @chance-coleman for the PR!

I have added two comments but I want that @mposolda also checks this. So do not start changing code until he also reviews the PR and gives his opinion.

slaskawi commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -96,6 +100,22 @@ public CloseableHttpClient getHttpClient() {
+            }
+
+            @Override
+            public CloseableHttpClient getRetriableHttpClient(RetryConfig retryConfig) {
```
@chance-coleman has a point here in my opinion. Depending on the area of the code, we may want to use different setting. In other words, what will work correctly for OCSP might not be a good fit for Identity Providers or other functionalities.

rmartinc replied:
Yep, this is just a matter of what we prefer. Let's see what other people think aboy this.

mposolda replied:
+1 for just `getRetriableHttpClient()` . If we ever need different configuration for different "use-cases" of HTTP client, we can add new method if needed.

I can imagine that we may want different configurations for different use-cases, however this might mean also different configurations for other aspects than just retry configuration. For more flexibility, maybe the best is to have the method like:
```
            @Override
            public CloseableHttpClient getHttpClient(String context) {
```
where context is for example something like "ocsp" or "facebook-idp" . Then in the configuration of Http-client provider, we can add namespaces for various configuration options. For example we can have `ocsp--proxy-mappings` if we want to have different proxy-mappings configuration for the case when HTTP client is called in the context of "ocsp"
. This will also allow to have different retry configurations for "ocsp" and different for other things. The `Config.Scope` has some support for namespaces in itself AFAIK.

I do not insist on this, but would prefer something like this instead of introducing `getRetriableHttpClient(RetryConfig)` , which IMO does not have much flexibility and at the same time, it is not used anywhere right now.

mposolda commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -165,88 +188,227 @@ public String getId() {
+                                            (Math.random() * retryConfig.getJitterFactor() * 2.0);
+                                    delay = (long)(baseDelay * jitter);
+                                }
+                                Thread.sleep(delay);
```
There is also `ServiceUnavailableRetryStrategy` . Could not that one be used instead of implementing the retry logic in our own codebase? For example see the 2nd post from https://stackoverflow.com/questions/48541329/timeout-between-request-retries-apache-httpclient .

mposolda commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -342,6 +504,57 @@ public List<ProviderConfigProperty> getConfigMetadata() {
                 .defaultValue(HttpClientProvider.DEFAULT_MAX_CONSUMED_RESPONSE_SIZE)
                 .add()
+                .property()
+                .name("http-client.default-max-retries")
```
I am not sure why those configuration options have prefix `http-client` here? Other config options of the provider do not have this prefix.

Also there is one confusing aspect: When someone reads the documentation of this provider, he would see `http-client-default-max-retries` configured to 3, which imply that this is used by default by HTTP client. But AFAIK, this is used just if `getRetriableHttpClient()` is called right? By default, the `getHttpClient()` or other methods do not use this retry configuration at all and requests are not retried. Which is not clear from this docs...

Might be better if we have something like `.name("default-max-retries")` (which would be 0 by default and hence disabled retries) and `.name("ocsp--default-max-retries")` (which would be 3 and hence enabled retries for OCSP by default). I proposed in the other comment to use "namespaces" for the configurations. This will be likely harder to implement, but seems to me like more clear configuration.

mposolda reviewed (commented):
@chance-coleman @rmartinc @slaskawi  Added some additional review comments. WDYT?

slaskawi commented:
Hey @mposolda 👋🏻 

Could I ask you to have a look at this one? It seems @chance-coleman addressed all the comments and I think we should be close to getting it in.

stianst reviewed (commented):
Not convinced about this should be configured on the X509 authenticator specifically. Looking at the issue is explicitly mentions OCSP. However, if network traffic is unstable then this will affect any outgoing HTTP requests from Keycloak.

I would argue we should instead consider having config options for this for the HTTP client provider in general.

One thing to bear in mind here though if there are a few retries, and the timeout is large, then the original incoming HTTP request to Keycloak would likely have timed out in the meantime. To me that is another argument to make it a server wide configuration option, as incoming request timeout should be higher than the total of retries + timeouts for outgoing requests.

@ahus1 @shawkins WDYT?

stianst reviewed (changes requested):
Thinking more about this, it really isn't something we should make configurable as fine-grained as within the X509 authenticator, but just make a server wide config option for outgoing http requests

chance-coleman committed:
Simplify HTTP client retry to server-wide configuration

Refactored retry behavior from per-feature configuration to a simpler
server-wide approach based on reviewer feedback. This eliminates UI
complexity and provides consistent retry behavior across all HTTP clients.

Key changes:
- Removed OCSP-specific retry configuration from X.509 authenticator UI
- Simplified HttpClientProvider interface (removed parameterized method)
- Consolidated retry configuration to DefaultHttpClientFactory
- Updated documentation to reflect server-wide configuration
- Renamed timeout properties to avoid conflicts with existing settings

Configuration is now opt-in (max-retries=0 by default) and applies to
all outgoing HTTP requests including OCSP validation, identity provider
communication, and other external calls.

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

slaskawi commented on server-spi-private/src/main/java/org/keycloak/connections/httpclient/HttpClientProvider.java:
```diff
@@ -40,6 +40,20 @@ public interface HttpClientProvider extends Provider {
+     * 
+     * @return A CloseableHttpClient with retry capabilities based on global configuration
+     */
+    default CloseableHttpClient getRetriableHttpClient() {
```
We can remove this method too.

slaskawi commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/AbstractX509ClientCertificateAuthenticatorFactory.java:
```diff
@@ -115,7 +115,6 @@ public abstract class AbstractX509ClientCertificateAuthenticatorFactory implemen
         serialnumberHex.setDefaultValue(Boolean.toString(false));
         serialnumberHex.setHelpText("Use the hex representation of the serial number. This option is relevant for authenticators using serial number.");
 
```
We can revert these empty line removals (and additions in other files) to avoid introducing noise to this PR.

slaskawi commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -165,88 +177,227 @@ public String getId() {
-                            throw new RuntimeException("Failed to load truststore", e);
+        // Create HTTP client builder
+        HttpClientBuilder builder = newHttpClientBuilder();
+
```
I wonder if we could revert the implementation to the original on, extract these settings into a separate method and call it like this:

```
RetryConfig retryConfig = ... ;
builder.retries(retryConfig);
```

All the wiring would go to the builder and the implementation would be very small and lean in this class. 

What do you think about this approach?

chance-coleman committed:
Simplify HTTP client retry to single client with server-wide configuration

Refactored HTTP client retry implementation based on reviewer feedback to
use a single HTTP client with built-in retry behavior instead of separate
retriable and non-retriable clients.

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

chance-coleman commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -68,13 +68,17 @@ public class DefaultHttpClientFactory implements HttpClientFactory {
     public static final String MAX_CONSUMED_RESPONSE_SIZE = "max-consumed-response-size";
 
     private volatile CloseableHttpClient httpClient;
+    private volatile CloseableHttpClient retriableHttpClient;
```
Removed retriableHttpClient field, built retry into the main httpClient

chance-coleman commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -165,88 +177,227 @@ public String getId() {
+
+        // Initialize server-wide retry configuration
+        int maxRetries = config.getInt("max-retries", 0); // No retries by default (opt-in)
+        boolean retryOnIOException = config.getBoolean("retry-on-io-exception", true);
```
Renamed property from retry-on-io-exception to retry-on-error

chance-coleman committed:
Simplify HTTP client retry to single client with server-wide configuration

Refactored HTTP client retry implementation based on reviewer feedback to
use a single HTTP client with built-in retry behavior instead of separate
retriable and non-retriable clients.

Key changes:
- Removed getRetriableHttpClient() method from HttpClientProvider interface
- Merged retry functionality into the main httpClient in DefaultHttpClientFactory
- Extracted retry configuration to configureRetries() helper method for cleaner code
- Renamed retry-on-io-exception property to retry-on-error (more generic)
- Removed per-feature retry configuration from X.509 authenticator
- Updated OCSPProvider to use standard getHttpClient() method
- Removed unused OCSP_CONNECT_TIMEOUT constant from OCSPProvider
- Removed retry-specific timeout properties (using standard socket-timeout-millis)

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

slaskawi commented on docs/documentation/server_admin/topics/authentication/x509.adoc:
```diff
@@ -143,6 +143,8 @@ Checks the certificate revocation status by using Online Certificate Status Prot
 *OCSP Fail-Open Behavior*::
 By default the OCSP check must return a positive response in order to continue with a successful authentication. Sometimes however this check can be inconclusive: for example, the OCSP server could be unreachable, overloaded, or the client certificate may not contain an OCSP responder URI. When this setting is turned ON, authentication will be denied only if an explicit negative response is recei ...
 
+OCSP retry behavior is configured server-wide through the HTTP client provider. See the Server Configuration Guide for details on configuring retry settings for all outgoing HTTP requests, including OCSP validation.
```
Could you please document the new settings in https://github.com/keycloak/keycloak/blob/37bea126c723068b35d1f46bab5367c18db156ab/docs/guides/server/outgoinghttp.adoc and add a reference here to that page?

slaskawi commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -230,8 +230,8 @@ private void lazyInit(KeycloakSession session) {
                     if (disableTrustManager) {
-                    	logger.warn("TrustManager is disabled");
-                    	builder.disableTrustManager();
+                    logger.warn("TrustManager is disabled");
```
Seems like a misaliged line.

slaskawi reviewed (commented):
The code LGTM! Requested only doc changes and a few nit fixes.

chance-coleman committed:
address feedback for more documentation and cleanup

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

chance-coleman committed:
fix whitespace

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

slaskawi reviewed (approved):
LGTM! Thanks for all the changes @chance-coleman !

@stianst This one is ready for re-review.

chance-coleman committed:
Merge branch 'main' into retry-httpclient-implementation

chance-coleman committed:
add doc callout about the danger of not configuring the timeout appropriately

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

stianst reviewed (commented):
Looks okay to me, haven't done a full review though

mposolda commented on server-spi-private/src/main/java/org/keycloak/utils/OCSPProvider.java:
```diff
@@ -126,13 +124,6 @@ protected byte[] getEncodedOCSPResponse(KeycloakSession session, byte[] encodedO
         HttpPost post = new HttpPost(responderUri);
         post.setHeader(HttpHeaders.CONTENT_TYPE, "application/ocsp-request");
-
-        final RequestConfig params = RequestConfig.custom()
```
This change makes sense, but strictly speaking, it is background incompatible change (as for example default socket timeout is 5 seconds for the default HTTP client provider).

IMO it is OK to include this change in your PR, but it would be also good to add a note about this to the file https://github.com/keycloak/keycloak/blob/main/docs/documentation/upgrading/topics/changes/changes-26_5_0.adoc . Probably to the "Notable changes" section. I assume that note can mention that OCSP behaviour used hardcoded socket and connection timeouts of 10 seconds before, but now it is updated to use the settings from the HTTP client configuration.

WDYT?

mposolda reviewed (commented):
@chance-coleman @slaskawi Thanks for all the updates! I've added few small inline comments. Could you please check them?

rmartinc reviewed (commented):
For me it's OK once the @mposolda comments are fixed. And I would use always true for the option `retry-on-error`, because I suppose we are always using common implementations (HttpGet, HttpPost,... that are not retried if the parameter is set to false).

mposolda commented on docs/guides/server/outgoinghttp.adoc:
```diff
@@ -57,6 +57,44 @@ Specify proxy configurations for outgoing HTTP requests. For more details, see <
+Maximum number of retry attempts for failed HTTP requests. Set to 0 to disable retries. Default: 0.
+
+*retry-on-error*::
+Whether to retry HTTP requests when errors occur. Default: true.
```
Yeah, I was able to see that. But without looking to the code, this option will not be clear to me at all just after looking at this docs :-)

Removing this option entirely from Keycloak config and always set to true when `max-retries >= 0` might be even better though?
@chance-coleman @slaskawi WDYT?

slaskawi replied:
@rmartinc has a good point. We could decide to turn the entire mechanism on/off by checking `max-retries`. I'm fully onboard with this idea.

chance-coleman committed:
refactor retry-on-error to only use max-retries, updated wildcard import, update changes doc for callout of ocsp retry behavior change

Signed-off-by: UnicornChance <chance@defenseunicorns.com>

chance-coleman committed:
Merge branch 'main' into retry-httpclient-implementation

Signed-off-by: Chance Coleman <139784371+chance-coleman@users.noreply.github.com>

mposolda reviewed (approved):
@chance-coleman @rmartinc @slaskawi Thanks for the additional updates, reviews and feedback. And thanks for your patience during the work on this PR!

I hope we can merge once tests are OK (Documentation external links tests are not related to this PR and are not a blocker for merging this).
